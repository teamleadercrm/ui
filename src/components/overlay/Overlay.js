import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';
import cx from 'classnames';
import theme from './theme.css';
import { KEY } from '../../constants';
import { selectOverlayNode } from '../select/Select';

class Overlay extends PureComponent {
  innerWrapperRef = React.createRef();

  componentDidMount() {
    const { active, lockScroll } = this.props;

    if (active && lockScroll) {
      document.body.style.overflow = 'hidden';
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.lockScroll) {
      const becomingActive = this.props.active && !prevProps.active;
      const becomingUnactive = !this.props.active && prevProps.active;

      if (becomingActive) {
        document.body.style.overflow = 'hidden';
      }

      if (becomingUnactive && !document.querySelectorAll('[data-teamleader-ui="overlay"]')[1]) {
        document.body.style.overflow = '';
      }
    }
  }

  componentWillUnmount() {
    if (this.props.active && this.props.lockScroll) {
      if (!document.querySelectorAll('[data-teamleader-ui="overlay"]')[1]) {
        document.body.style.overflow = '';
      }
    }
  }

  handleEscKey = (event) => {
    if (this.props.active && event.key === KEY.Escape) {
      event.stopPropagation();
      // react-select has its own implementation of an overlay, conflicting with our custom implementation
      // so escape events fired while the select menu is open should be ignored
      const selectMenuOpen = selectOverlayNode.hasChildNodes();
      !selectMenuOpen && this.props.onEscKeyDown && this.props.onEscKeyDown(event);
    }
  };

  handleClick = (event) => {
    event.stopPropagation();
    // Only register clicks outside of the children
    if (
      this.props.onClick &&
      !this.innerWrapperRef.current?.contains(event.target) &&
      // react-select has its own implementation of an overlay, conflicting with our custom implementation
      // so clicks on the select overlay shouldn't be registered either
      !selectOverlayNode.contains(event.target)
    ) {
      this.props.onClick(event);
    }
  };

  render() {
    const { active, className, backdrop, lockScroll, onEscKeyDown, ...other } = this.props;

    return (
      <Transition timeout={0} in={active} appear>
        {(state) => {
          return (
            <div
              data-teamleader-ui="overlay"
              {...other}
              onKeyDown={this.handleEscKey}
              onClick={this.handleClick}
              className={cx(
                theme['overlay'],
                theme[backdrop],
                {
                  [theme['is-entered']]: state === 'entered',
                },
                className,
              )}
            >
              <div ref={this.innerWrapperRef}>{this.props.children}</div>
            </div>
          );
        }}
      </Transition>
    );
  }
}

Overlay.propTypes = {
  active: PropTypes.bool,
  backdrop: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  lockScroll: PropTypes.bool,
  onClick: PropTypes.func,
  onEscKeyDown: PropTypes.func,
};

Overlay.defaultProps = {
  backdrop: 'dark',
  lockScroll: true,
};

export default Overlay;
