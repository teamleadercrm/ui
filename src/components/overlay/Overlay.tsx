import React, { ReactNode, useRef, KeyboardEvent, MouseEvent, useEffect } from 'react';
import Transition from 'react-transition-group/Transition';
import cx from 'classnames';
import theme from './theme.css';
import { KEY } from '../../constants';
import { selectOverlayNode } from '../select/Select';
import { GenericComponent } from '../../@types/types';
import { BoxProps } from '../box/Box';

export interface OverlayProps extends Omit<BoxProps, 'className' | 'children'> {
  active?: boolean;
  backdrop?: string;
  children?: ReactNode;
  className?: string;
  lockScroll?: boolean;
  onEscKeyDown?: (event: KeyboardEvent) => void;
  onOverlayClick?: (event: MouseEvent) => void;
}

export const Overlay: GenericComponent<OverlayProps> = ({
  active,
  backdrop = 'dark',
  children,
  className,
  lockScroll = true,
  onEscKeyDown,
  onOverlayClick,
  ...other
}) => {
  const innerWrapperRef = useRef<HTMLDivElement | null>(null);
  const clickOriginRef = useRef<Node | null>(null);

  const handleEscKey = (event: KeyboardEvent) => {
    if (active && event.key === KEY.Escape) {
      event.stopPropagation();
      // react-select has its own implementation of an overlay, conflicting with our custom implementation
      // so escape events fired while the select menu is open should be ignored
      const selectMenuOpen = selectOverlayNode.hasChildNodes();
      !selectMenuOpen && onEscKeyDown && onEscKeyDown(event);
    }
  };

  const handleMouseDown = (event: MouseEvent) => {
    clickOriginRef.current = event.target as Node;
  };

  const handleMouseUp = (event: MouseEvent) => {
    // Only register clicks outside of the children
    if (
      onOverlayClick &&
      // if the clickOrigin is no longer part of the DOM tree, (f.e. due to internal React re-renders)
      document.body.contains(clickOriginRef.current) &&
      !innerWrapperRef.current?.contains(clickOriginRef.current) &&
      // react-select has its own implementation of an overlay, conflicting with our custom implementation
      // so clicks on the select overlay shouldn't be registered either
      !selectOverlayNode.contains(clickOriginRef.current) &&
      // make sure only clicks within the current portal's DOM tree are handled
      event.currentTarget.contains(event.target as Node)
    ) {
      onOverlayClick(event);
    }
  };

  useEffect(() => {
    if (lockScroll) {
      if (active) {
        document.body.style.overflow = 'hidden';
      }

      if (!active && !document.querySelectorAll('[data-teamleader-ui="overlay"]')[1]) {
        document.body.style.overflow = '';
      }
    }
  }, [active, lockScroll]);

  return (
    <Transition timeout={0} in={active} appear>
      {(state) => {
        return (
          <div
            data-teamleader-ui="overlay"
            {...other}
            onKeyDown={handleEscKey}
            onMouseDown={handleMouseDown}
            onMouseUp={handleMouseUp}
            className={cx(
              theme['overlay'],
              theme[backdrop],
              {
                [theme['is-entered']]: state === 'entered',
              },
              className,
            )}
          >
            <div ref={innerWrapperRef}>{children}</div>
          </div>
        );
      }}
    </Transition>
  );
};

export default Overlay;
