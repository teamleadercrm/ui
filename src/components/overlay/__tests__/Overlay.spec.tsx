import React from 'react';
import { render } from '@testing-library/react';
import Overlay from '../Overlay';

describe('Component - Overlay', () => {
  it('renders', async () => {
    const { asFragment } = render(<Overlay>The children</Overlay>);
    expect(asFragment()).toMatchSnapshot();
  });
});
