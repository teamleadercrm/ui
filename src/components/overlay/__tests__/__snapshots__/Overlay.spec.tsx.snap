// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Overlay renders 1`] = `
<DocumentFragment>
  <div
    class="overlay dark"
    data-teamleader-ui="overlay"
  >
    <div>
      The children
    </div>
  </div>
</DocumentFragment>
`;
